File: receiver_py/enlace_crc32.py
# Capa Enlace - CRC32 (Receptor) 
from typing import Tuple
POLY = 0x04C11DB7

def crc32_msb_bits(bits: str) -> int:
    reg = 0xFFFFFFFF
    for c in bits:
        bit = 1 if c == "1" else 0
        feedback = ((reg >> 31) & 1) ^ bit
        reg = ((reg << 1) & 0xFFFFFFFF)
        if feedback:
            reg ^= POLY
    return reg ^ 0xFFFFFFFF

def to_bits32_msb(x: int) -> str:
    return "".join("1" if (x >> i) & 1 else "0" for i in range(31,-1,-1))

def verificar(frame_bits: str) -> Tuple[str, str]:
    if len(frame_bits) < 33:
        print("[ENLACE][RECEPTOR][CRC32] Trama inválida (<33 bits).")
        return "DISCARD", ""
    payload, recv_crc_bits = frame_bits[:-32], frame_bits[-32:]
    calc_crc_bits = to_bits32_msb(crc32_msb_bits(payload))
    print(f"[ENLACE][RECEPTOR][CRC32] payload_bits: {payload}")
    print(f"[ENLACE][RECEPTOR][CRC32] CRC recibido:   {recv_crc_bits}")
    print(f"[ENLACE][RECEPTOR][CRC32] CRC calculado: {calc_crc_bits}")
    if recv_crc_bits == calc_crc_bits:
        print("[ENLACE][RECEPTOR][CRC32] Verificación: OK (sin errores).")
        return "OK", payload
    else:
        print("[ENLACE][RECEPTOR][CRC32] Verificación: ERROR -> DESCARTAR.")
        return "DISCARD", ""

File: emitter_cpp/src/enlace_crc32.cpp
#include "enlace_crc32.hpp"
#include <string>
#include <cstdint>

static uint32_t crc32_msb_bits(const std::string& bits_msb){
    const uint32_t POLY = 0x04C11DB7u;
    uint32_t reg = 0xFFFFFFFFu;
    for(char c : bits_msb){
        uint32_t bit = (c=='1') ? 1u : 0u;
        uint32_t feedback = ((reg >> 31) & 1u) ^ bit;
        reg = (reg << 1) & 0xFFFFFFFFu;
        if(feedback) reg ^= POLY;
    }
    return reg ^ 0xFFFFFFFFu;
}

static std::string to_bits32_msb(uint32_t x){
    std::string s; s.reserve(32);
    for(int i=31;i>=0;i--) s.push_back( ((x>>i)&1u) ? '1':'0' );
    return s;
}

namespace enlace_crc32 {
    std::string append_crc32(const std::string& payload_bits) {
        uint32_t crc = crc32_msb_bits(payload_bits);
        std::string crc_bits = to_bits32_msb(crc);
        return payload_bits + crc_bits;
    }

    std::string crc32_bits_msb(const std::string& payload_bits) {
        uint32_t crc = crc32_msb_bits(payload_bits);  
        return to_bits32_msb(crc);                   
    }
}
